magma/passes/clock.py

```diff
@@ -1,10 +1,201 @@
+    """
+    default_clocks = {}
+    for clock_type in ClockTypes:
+        clocks = get_output_clocks_in_defn(defn, clock_type)
+        try:
+            clock = only(clocks)
+        except IterableException:
+            clock = None
+        default_clocks[clock_type] = clock
+    return default_clocks
+
+
+def _get_undriven_clocks_in_array(
+        value: Array, clock_type: Kind) -> Iterable[Type]:
+    """
+    Returns all undriven values of type @clock_type contained in array @value.
+    """
+    # First check if the first element has any undriven.
+    it = get_undriven_clocks_in_value(value[0], clock_type)
+    try:
+        first = next(it)
+    except StopIteration:  # early exit to avoid traversing other elements
+        return
+    yield first
+    yield from it
+    # NOTE(leonardt): Since magma arrays don't support zero-length values, we
+    # have to check the size before slicing.
+    if len(value) <= 1:
+        return
+    for elem in value[1:]:
+        yield from get_undriven_clocks_in_value(elem, clock_type)
+
+
+def get_undriven_clocks_in_value(
+        value: Type, clock_type: Kind) -> Iterable[Type]:
+    """Returns all undriven values of type @clock_type contained in @value."""
+    if isinstance(value, Tuple):
+        for elem in value:
+            yield from get_undriven_clocks_in_value(elem, clock_type)
+        return
+    if isinstance(value, Array):
+        yield from _get_undriven_clocks_in_array(value, clock_type)
+        return
+    if isinstance(value, clock_type) and value.trace() is None:
+        # Trace back to last undriven driver.
+        while value.driven():
+            value = value.value()
+        yield value
+    return
+
+
+def _drive_value_with_clock(value: Type, clk: Type):
+    _logger.info(f"Auto-wiring {repr(clk)} to {repr(value)}")
+    value @= clk
+
+
+def drive_undriven_clocks_in_inst(
+        defn: DefineCircuitKind, inst: Circuit, clock_type: Kind):
+    """
+    Drives all undriven input ports of @inst of type @clock_type with a default
+    driver of type @clock_type contained in @defn. If no such default driver
+    exists (either due to no drivers or multiple drivers existing), then this
+    function is a no-op.
+    """
+    undrivens = []
+    for port in inst.interface.inputs(include_clocks=True):
+        undrivens = itertools.chain(
+            undrivens, get_undriven_clocks_in_value(port, clock_type))
+    try:
+        undriven = next(undrivens)
+    except StopIteration:
+        return
+    clks = get_output_clocks_in_defn(defn, clock_type)
+    try:
+        clk = only(clks)
+    except EmptyIterableException:
+        _logger.warning(
+            f"Found no clocks in {defn.name}; skipping auto-wiring "
+            f"{clock_type}")
+        return
+    except NonSignletonIterableException:
+        _logger.warning(
+            f"Found multiple clocks in {defn.name}; skipping auto-wiring "
+            f"{clock_type}")
+        return
```

Comment:
Could this be done at the top level to avoid having to traverse the instances?  (i.e. check once if there's a default clock, otherwise skip)  Also we shouldn't need to find these for every instance (instead reuse them across all instances)

Reply:
We could run a pass that does get_output_clocks_in_defn and stores it in the defn object

Reply:
I guess there's a small tradeoff with the common case where clocks are manually wired, so we don't always need to store this information.  Another option is to run this logic once and cache it for future needs (so it's only run on demand when we encounter an undriven clock, but not run more than once)

Reply:
Yeah that all makes sense. The reason I did it in this order was for diagnostics. If there are no undriven clocks in the defn then it doesn't matter whether or not there is a default clock. So we need to do that check first before checking for default clocks.

We could first check for default clocks, and then check for undriven clocks but that makes the code pretty ugly and results in some duplicated code -- but I do agree this is a perf. tradeoff.

My longer term plan is to make clocks more of a first class citizen. That is, when a definition is created we stash some information about it's clocks explicitly and then have a well-defined API to query about those clocks. That will also be useful when thinking about multiple clock domains etc. So as long as we don't observe a performance regression with this code, my plan is to forego more work on this until we make this change.